HCollections/HListT.fs

```diff
@@ -25,96 +25,72 @@ module HListT =
 
     let toTypeList<'ts, 'elem> (list : HListT<'ts, 'elem>) : 'ts TypeList =
         match list with
-        | Empty teq ->
-            TypeList.empty
-            |> Teq.castFrom (TypeList.cong teq)
-        | Cons (_, tl) -> tl
+        | Empty (h, _)
+        | Cons (h, _, _) -> HList.toTypeList h
 
-    let empty<'elem> : HListT<unit, 'elem> = HListT.Empty Teq.refl
+    let toHList<'ts, 'elem> (input : HListT<'ts, 'elem>) : 'ts HList =
+        match input with
+        | Empty (hlist, _)
+        | Cons (hlist, _, _) -> hlist
+
+    let toList<'ts, 'elem> (input : HListT<'ts, 'elem>) : 'elem list =
+        match input with
+        | Empty _ -> []
+        | Cons (_, elems, _) -> elems
+
+    let empty<'elem> : HListT<unit, 'elem> = HListT.Empty (HList.empty, Teq.refl)
 
     let length<'ts, 'elem> (xs : HListT<'ts, 'elem>) : int =
         match xs with
         | Empty _ -> 0
-        | Cons (_, tl) -> TypeList.length tl
+        | Cons (_, elems, _) -> List.length elems
 
     let cons<'t, 'ts, 'elem> (x : 't) (elem : 'elem) (xs : HListT<'ts, 'elem>) =
-        let crate =
-            { new HListTConsCrate<_, _> with
-                member __.Apply e = e.Eval x elem xs Teq.refl
-            }
-        let tl =
-            xs
-            |> toTypeList
-            |> TypeList.cons<'t, _>
-
-        HListT.Cons (crate, tl)
+        match xs with
```

Comment:
Yeah, this code got more confusing because of the change to the structure of the crate, I think.

Reply:
I think this is just to do with me rushing to get this implementation done. I've taken another look at it and have rewritten it

Reply:
One small tweak I'd suggest here - I think you can replace all 4 lines (the match and pipe into the function) with:
`HListT.Cons (HList.cons x (toHList xs), elem::(toList xs), cons)`
That should have the same performance as the current implementation and actually saves on allocating the pair

Reply:
Might be a micro-optimisation but we could reduce the number of match statements by also having the empty case store an empty elem list